File: 프로그래머스/unrated/159993. 미로 탈출/미로 탈출.py
def BFS(sx, sy, x, y, maps):
    check = list([0] * len(maps[i]) for i in range(len(maps)))
    que = [[sx,sy]]
    dx = [0,1,0,-1]
    dy = [1,0,-1,0]
    flag = False
    
    while len(que):
        pos = que[0]
        que.pop(0)
        if pos == [x, y]:
            flag = True
            break
        
        for i,j in zip(dx, dy):
            nx = pos[0] + i
            ny = pos[1] + j
            if 0 <= nx < len(maps) and 0 <= ny < len(maps[0]):
                if maps[nx][ny] != 'X' and check[nx][ny] == 0:
                    check[nx][ny] = check[pos[0]][pos[1]] + 1
                    que.append([nx, ny])
                    
    return check[x][y] if flag else -1

def solution(maps):
    answer = 0
    start = []
    lever = []
    exit_ = []
    for i in range(len(maps)):
        if 'S' in maps[i]:
            start.append(i)
            start.append(maps[i].index('S'))
        if 'E' in maps[i]:
            exit_.append(i)
            exit_.append(maps[i].index('E'))
        if 'L' in maps[i]:
            lever.append(i)
            lever.append(maps[i].index('L'))
    
    #print(start)
    to_lev = BFS(start[0], start[1], lever[0], lever[1], maps)
    to_exi = BFS(lever[0], lever[1], exit_[0], exit_[1], maps)

    return -1 if(to_lev == -1 or to_exi == -1) else to_lev + to_exi

File: 프로그래머스/unrated/159993. 미로 탈출/미로 탈출.cpp
#include <bits/stdc++.h>
using namespace std;
class point{
public:
    int x_ = 0 ;
    int y_ = 0 ;
    
    point(int x, int y){x_ = x, y_ = y;}
    point(){x_ = 0, y_ = 0;}
};

int dx[] = {0, 1, 0, -1} ;
int dy[] = {1, 0, -1, 0} ;

int BFS(point s, point e, vector<string> maps) {
    queue<point> que ;
    int chk[maps.size()][maps[0].size()];
    memset(chk, 0, sizeof(chk));
    chk[s.x_][s.y_] = 1 ;
    
    que.push(s) ;
    int cnt = 0 ;
    while(!que.empty()) {
        int size = que.size() ;
        for(int i = 0 ; i < size ; i++) {
            point now = que.front() ; que.pop() ;
            for(int j = 0 ; j < 4 ; j++) {
                int nx = now.x_ + dx[j] ;
                int ny = now.y_ + dy[j] ;
                if(nx < 0 || nx >= maps.size() || ny < 0 || ny >= maps[nx].size()) continue ;
                if(chk[nx][ny] || maps[nx][ny] == 'X') continue ;
                
                chk[nx][ny] = chk[now.x_][now.y_] + 1 ;
                que.push(point(nx, ny)) ;
                
                if(nx == e.x_ && ny == e.y_) return chk[nx][ny] ;
            }
        }
    }
    return -1 ;
}

int solution(vector<string> maps) {
    int answer = 0;

    point S = point();
    point E = point();
    point L = point();
        for(int i = 0 ; i < maps.size() ; i++)
            for(int j = 0 ; j < maps[i].size() ; j++)
                switch (maps[i][j]) {
                    case 'S': S.x_ = i, S.y_ = j ; break ;
                    case 'E': E.x_ = i, E.y_ = j ; break ;
                    case 'L': L.x_ = i, L.y_ = j ; break ;
                }
    int StoL = BFS(S, L, maps) - 1 ;
    int LtoE = BFS(L, E, maps) - 1 ;
    return (StoL >= 0 && LtoE >= 0) ? StoL + LtoE : -1;
}